### PR TITLE
Allow conversion of `std::process::Command` into a `openssh::Command` given a `openssh::Session`.

### PR DESCRIPTION
**Pull request by aalekhpatel07**
The `std::process::Command` and the `openssh::Command` APIs are fairly similar, so I wonder if it is reasonable to allow an arbitrary `std::process::Command` to be run over ssh.

#### Example

```rust
use std::process;
use openssh::{OverSsh, Session, KnownHosts};

#[tokio::main]
pub async fn main() {
    let session = Session::connect_mux("me@ssh.example.com", KnownHosts::Strict).await.unwrap();

    let ls = process::Command::new("ls")
        .arg("-l")
        .arg("-h")
        .over_ssh(&session)
        .expect("no env vars or current_dir are specified")
        .output()
        .await
        .unwrap();

    println!("{}", String::from_utf8_lossy(&ls.stdout));
}


```

### CODE REVIEW

**Comment by jonhoo** [comment]
This change is [Reviewable](https://reviewable.io/reviews/openssh-rust/openssh/112)

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:59]: We could also implement this for `tokio::process::Command`.

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:58]: I think it will be also great to have some stub implementation for `&T`, `Box<T>`, `Arc<T>` and `Rc<T>` like this:

```
impl<T: OverSsh> OverSsh for &T {
    fn over_session<'session>(&self, session: &'session Session) -> crate::Command<'session> {
        <T as OverSsh>::over_session(session)
    }
}
```

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:82]: I think we should use `Session::command` here to escape the program, since the ssh protocol essentially concats program and args into one string `$program $arg0 $arg1 ...` and pass it to shell.

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:83]: Same here, I think `Command::args` is more appropriate here.

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:61]: We should also document that this does not include any environment variable and the current dir.

**Review by NobodyXu — CHANGES_REQUESTED** [review]
Thanks!

I really like the design and the implementation is simple, but there are a few places where it can be improved.

**Comment by aalekhpatel07** [comment]
wow thank you for your kind words and a quick review. I'll clean it up right away! Thanks for your time.

File: src/command.rs
Comment (NobodyXu) [reply src/command.rs:58]: For `Box<T>`, `Arc<T>` and `Rc<T>`, we could perhaps use [`AsRef`](https://doc.rust-lang.org/std/convert/trait.AsRef.html) trait, which is more extensible and will work for types defined in other crates, such as `triomphe`'s `Arc`, though I'm really not sure whether this will work, so it would need some testing:

```
impl<T: AsRef<U>, U: OverSsh> OverSsh for T {
    fn over_session<'session>(&self, session: &'session Session) -> crate::Command<'session> {
        self.as_ref().over_session(session)
    }
}
```

For `&T`, we might still need to manually implement for them.

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:65]:
```suggestion
    /// # #[tokio::main(flavor = "current_thread")]
    /// # async fn main() -> Result<(), Box<dyn Error>> {
```

Comment (NobodyXu) [reply src/command.rs:65]: Inside the code block, anything starts with `#` will be hidden, so you can use `#[tokio::main]` to fix the error here.

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:78]:
```suggestion
    /// #    Ok(())
    /// # }
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/command.rs
Comment (aalekhpatel07) [reply src/command.rs:82]: hmm I'm not sure I fully understand this. As it stands, `self.get_program()` returns an `&OsStr` but `Session::command` expects `Into<Cow<'a, str>>`. A meaningful way to get that is to `self.get_program().to_string_lossy()` but I'm a little worried that changes the semantics here. please lmk if I misunderstood it.

Comment (NobodyXu) [reply src/command.rs:82]: Oh I didn't realize the difference between typing.

Yeah that's certainly an issue, but I still think it's necessary to escape them, e.g. if the program contains `'` or `!`, then it will be interpreted by the shell instead of as part of the program name.

We could use [`OsStrExt::as_bytes`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html#tymethod.as_bytes) then escapes them in the same way as [`shell_escape::unix::escape`](https://docs.rs/shell-escape/latest/src/shell_escape/lib.rs.html#101-120).

File: src/command.rs
Comment (aalekhpatel07) [reply src/command.rs:58]: No dice with the type checker. Seems like `U` isn't sufficiently constrained. Even this doesn't work:

```rust
impl<T: Into<U>, U: Foo> Foo for T {
    ...
}
```

I feel like I've tried to write something similar before and come across it. I'm not sure I know what its called though.

File: src/command.rs
Comment (aalekhpatel07) [reply src/command.rs:82]: There are still some typing incompatibilities as `shell_escape::unix::escape` expects a `Cow<str>`. I tried to implement a variant for `Cow<OsString> -> Cow<OsString>` but I couldn't find a way to interleave special (i.e. `[^\!']`) characters with the non-special ones. 

Hoping you could suggest a fix for what you have in mind.

File: src/command.rs
Comment (aalekhpatel07) [reply src/command.rs:83]: same story as with `Session::command` (i.e. expected `Cow<str>` but provided `Cow<OsStr>`).

File: src/command.rs
Comment (NobodyXu) [reply src/command.rs:62]: Well, that's not unexpected...

For now let's go with manually impl for `&T`, `Box<T>`, `Arc<T>` and `Rc<T>`

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:74]:
```suggestion
    ///
```

Having a blank line like this would not show up in the example on docs.rs

It has to be prefixed with `///`.

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:148]: I don't think we need this as `impl<S> OverSsh for &S` should be enough and `rustc` has auto deref rules that should take care for this.

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:124]: Let's move the doc to the definition of `OverSsh` trait, since users will check it first and it will avoid the dup doc for both impl.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/command.rs
Comment (NobodyXu) [reply src/command.rs]: You can use [`OsStrExt::as_bytes`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html#tymethod.as_bytes) to convert `OsStr` to `&[u8]` first, then you can implement [`shell_escape::unix::escape`](https://docs.rs/shell-escape/latest/src/shell_escape/lib.rs.html#101-120) but for bytes `&[u8]` instead of `&str`.

Comment (NobodyXu) [reply src/command.rs]: I just recall there is plan to support windows, but it will take probably at least another 3-6 months before it is even ready for review (I am working some other projects right now).

So this is probably fine.
For windows, we would have to use some different implementations.

**Comment by NobodyXu** [comment]
Ehhh we now have a rustc ICE.

@aalekhpatel07 Would you like to open an issue in rust-lang/rust?

**Comment by aalekhpatel07** [comment]
> Ehhh we now have a rustc ICE.
> 
> @aalekhpatel07 Would you like to open an issue in rust-lang/rust?

Haha for sure. Any preference for a name?

**Comment by NobodyXu** [comment]
@aalekhpatel07 Looks like some of the ICEs happens in mir borrow checking, so perhaps "mir borrow checking ICE in openssh-rust" will be a good name.

Also I recommend you to create a separate branch, containing the commit causing the ICEs just so it's easier for others to reproduce the ICEs.

File: src/command.rs
Comment (aalekhpatel07) [reply src/command.rs]: that makes sense. lets go with this for now.

**Comment by aalekhpatel07** [comment]
I believe the new `escape(s: &[u8]) -> String` should do the job for us. lmk

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:101]:
```suggestion
        let args = self
            .get_args()
            .map(|arg| {
                let arg_escaped = escape(arg.as_bytes());
                Cow::Borrowed(arg_escaped.as_str())
            });
        command.args(args);
```

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:23]: Let's make this easier to use and easier to port to windows by accept `&OsStr`.

Also, I'd like the return type to be `Cow<'_, OsStr>`, not `String`.

Otherwise what's the point of using `OsStr`?
We could just use `OsStr::to_string_lossy`.

```suggestion
pub fn escape(s: &OsStr) -> Cow<'_, OsStr> {
```

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:29]: This is just wrong, there's no guarantee that `OsStr` contains utf-8 characters.

File: src/lib.rs
Comment (NobodyXu) [inline src/lib.rs:173]: No, let's not export this.
It should not be part of the public API.

Comment (NobodyXu) [reply src/lib.rs:173]: Actually, I'd like this to be a separate crate, but that is out of the scope of this PR.

Let's just make this crate-internal API.

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:8]: Let's take byte here:

```suggestion
fn whitelisted(byte: u8) -> bool {
```

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:32]: Same here, should use `Vec` and then use [`std::os::unix::ffi::OsStringExt::from_vec`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStringExt.html#tymethod.from_vec) to convert it to `OsStr`.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

**Comment by NobodyXu** [comment]
> (I'm sorry I think I accidentally removed the review request for jonhoo).

No worries, I've re-requested it.

**Comment by NobodyXu** [comment]
Aha, the CI reliably failed with ICEs from `rustc`.

File: src/escape.rs
Comment (aalekhpatel07) [reply src/escape.rs:29]: that's my bad. I was mistaken in assuming that each of those bytes individually representing whitelisted chars implies they are to be interpreted as single byte utf-8 chars.

**Comment by aalekhpatel07** [comment]
okay how about these changes? (thank you for being patient.)

File: src/command.rs
Comment (NobodyXu) [inline src/command.rs:97]:
```suggestion
            .map(escape);
```

File: src/lib.rs
Comment (NobodyXu) [inline src/lib.rs:172]:
```suggestion
mod escape;
```

`lib.rs` is the top-level of a crate, any definition here we be available to all submodules.

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:61]: It would be great if we can also test some non-utf8 characters, but I don't consider it necessary for this PR to be merged.

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:32]:
```suggestion
    let all_whitelisted = s.iter().copied().all(whitelisted);
```

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:44]: Nit: I think we can `escaped.reserve(4)` here to reduce allocation.

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:35]: You can use `Cow::Borrowed` on the original `s: &OsStr` here.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

**Comment by NobodyXu** [comment]
> okay how about these changes? (thank you for being patient.)

It's much better, but we would still need @jonhoo 's approval on this.
I'm not entirely sure about escaping since the utf encoding is a mess, so I just want to be safe and ask @jonhoo to also review this.

**Comment by NobodyXu** [comment]
@aalekhpatel07 BTW, don't forget to create a new issue on rust-lang/rust and linked it to this.
We would also have to waiting for the compiler guys to investigate and fix the issues.

**Comment by NobodyXu** [comment]
@aalekhpatel07 Hmmm, let me create the bug report for this.

**Comment by NobodyXu** [comment]
@aalekhpatel07 I've created an ICE bug report on rust-lang/rust.

Note that the lint CI did work as intended and you need to fix all `cargo-clippy` warnings.

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:35]: Ehh, I don't mean to reserve it here.

```suggestion
```

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:41]: I'd want you to do the reserve here in the match, not outside the loop:

```suggestion
                escaped.reserve(4);
                escaped.push(b'\'');
```

File: src/escape.rs
Comment (NobodyXu) [inline src/escape.rs:61]:
```suggestion
        test_escape_from_bytes(input.as_bytes(), expected.as_bytes())
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]
Almost done from me, some small changes requested...

Regarding the ICEs, there's already a PR in rust-lang/rust to fix this.

**Review by NobodyXu — APPROVED** [review]
Now we just need to wait for @jonhoo to review this and upstream to fix the ICEs on stable channel.

**Comment by aalekhpatel07** [comment]
They have a PR for that ICE already??

Wow! I'm blown away by how responsive, knowledgeable, and supportive the contributors are in the Rust ecosystem. 

Thank you for all your help, @NobodyXu, for helping me improve the PR and develop my understanding of `Cow`s and `OsStr`s in Rust.

**Comment by NobodyXu** [comment]
> They have a PR for that ICE already??

Yes, https://github.com/rust-lang/rust/pull/107532

Though I have no idea on whether it will land in 1.68 or will they have a backport it to 1.67.1 since whether a backport will happen depends on the severity of the bug.

> Thank you for all your help, @NobodyXu, for helping me improve the PR and develop my understanding of `Cow`s and `OsStr`s in Rust.

No worries!

File: src/command.rs
Comment (jonhoo) [inline src/command.rs:87]: It's generally common in Rust to have the method of a single-purpose trait match the name of that trait. See for example `Clone::clone`, `AsRef::as_ref`, `Hash::hash`, etc. So we should either name this trait `OverSession` or name the method `over_ssh`. I _think_ I prefer the latter, because it's clearer what's going on from reading just the code. We _could_ also make it `OverSsh::over_ssh_session`, though the verbosity is a little unfortunate.

File: src/command.rs
Comment (jonhoo) [inline src/command.rs:61]: What does "set explicitly" mean here?

Comment (jonhoo) [reply src/command.rs:61]: Also, I wonder if there's an argument for us simply returning an error if we find that the current working directory or an environment variable is overridden?

Alternatively, can we transparently handle that as well? I guess probably not without assuming the remote end is Bash?

File: src/command.rs
Comment (jonhoo) [inline src/command.rs:55]:
```suggestion
/// If a command is `OverSsh` then it can be executed over an SSH session.
///
/// Primarily a way to allow `std::process::Command` to be turned directly into an `openssh::Command`.
```
Entirely a nit: it will make the rustdoc render a little nicer.

File: src/command.rs
Comment (jonhoo) [inline src/command.rs:96]: Why use `escape` + `raw_args` rather than just calling `args` directly and have that do the escaping for us?

Comment (jonhoo) [reply src/command.rs:96]: Same above for `escape` + `raw_command`

File: src/command.rs
Comment (jonhoo) [inline src/command.rs:142]: I think these probably aren't worth including. Even the impl for `Box` feels a little odd. I also think `impl for &mut` is missing, if anything (but again, I think just having the `impl for &` is probably good enough).

File: src/escape.rs
Comment (jonhoo) [inline src/escape.rs:2]: Can we submit a PR to `shell-escape` to add a version supporting `OsStr` there directly? I'm okay with us landing our own version initially, but it'd be way better if we could just use a version from them in the future.

File: src/escape.rs
Comment (jonhoo) [inline src/escape.rs:14]: Super nit, but let's call this "allowed".

File: tests/openssh.rs
Comment (jonhoo) [inline tests/openssh.rs:297]: I'd like to see this also test some cases that require escaping.

**Review by jonhoo — CHANGES_REQUESTED** [review]
This is a really neat idea! Thank you both for pushing it forward. I've left some notes.

File: src/command.rs
Comment (NobodyXu) [reply src/command.rs:96]: Because command/args only takes str and here we have OsStr.

Once this PR is merged, I plan to use escape introduced here for OsStr inside command/args so that they can take OsStr.

File: src/command.rs
Comment (NobodyXu) [reply src/command.rs:142]: I think impl for &T can also handle &mut T given rust auto-deref rules?

File: src/command.rs
Comment (NobodyXu) [reply src/command.rs:61]: The mux protocol actually supports env specification, though I didn't expose that because openssh didn't have that method.

For setting current dir, AFAIK, ssh does not support that.

Comment (jonhoo) [reply src/command.rs:61]: In that case, we should error (if possible) if `Command::current_dir` or `Command::env` has been called on the `Command` we're passed in.

File: src/command.rs
Comment (jonhoo) [reply src/command.rs:142]: I think that's only sort of true. In particular it still means that `&mut OverSssh` doesn't implement `OpenSsh`, which can be annoying in certain cases.

File: src/escape.rs
Comment (aalekhpatel07) [reply src/escape.rs:2]: Great suggestion! I've initiated [a PR](https://github.com/sfackler/shell-escape/pull/9) to `shell-escape` and if it gets merged, we can avoid having a `src/escape.rs` altogether.

File: src/command.rs
Comment (aalekhpatel07) [reply src/command.rs:87]: I do believe `OverSsh::over_ssh` makes it quite a bit clearer. That's an excellent improvement. Thank you.

File: tests/openssh.rs
Comment (aalekhpatel07) [reply tests/openssh.rs]: I'm not quite sure what a good test case should look like here and I'd greatly appreciate a hint. :thinking:

**Comment by aalekhpatel07** [comment]
one more set of changes as suggested in the review. please lmk if I missed something or misinterpreted a suggestion.

File: tests/openssh.rs
Comment (jonhoo) [reply tests/openssh.rs]: Ah, so, I was thinking things like having an argument that contains, say, `'`, ` `, and `\`.

File: src/command.rs
Comment (jonhoo) [inline src/command.rs:106]: This doesn't seem right — did you mean `.env`?

File: src/command.rs
Comment (jonhoo) [inline src/command.rs:108]: This doesn't currently compile.

**Review by jonhoo — CHANGES_REQUESTED** [review]

File: src/command.rs
Comment (aalekhpatel07) [reply src/command.rs:106]: You're absolutely right. That is an oversight on my part. Should've been `Command::new("echo").env("MY_ENV_VAR", "foo")`

File: src/command.rs
Comment (aalekhpatel07) [reply src/command.rs:108]: hmm. after the `.env` fix, maybe we should use `assert!(matches!(...))` instead.

File: tests/openssh.rs
Comment (jonhoo) [inline tests/openssh.rs:326]: I'd like to see a space (` `) in here too :)

**Review by jonhoo — APPROVED** [review]
Thanks for sticking with this!

**Comment by jonhoo** [comment]
Looks like rustfmt is complaining, and somehow we're also hitting a compiler ICE?
https://github.com/openssh-rust/openssh/actions/runs/4153289081/jobs/7184837200

**Comment by aalekhpatel07** [comment]
Really appreciate yours and @NobodyXu's  precious time reviewing this! I was able to learn a few things from this review, so thank you! Now hopefully the other PR in `shell-escape` also lands soon.

**Comment by aalekhpatel07** [comment]
yep the compiler ICE was pointed out by @NobodyXu last week and I believe there was some work happening in rust-lang/rust#107532 to fix it. Rustfmt should be happy now though.

**Comment by jonhoo** [comment]
I'll hold off on merging this until the fix for the ICE lands then 👍 Nothing more needed from your end at this point :)

**Comment by NobodyXu** [comment]
Since the ICE is fixed and the CI has passed, I will merge this PR.

**Pull request merged**
